sdk/core/azure-core-serializer-avro-apache/src/main/java/com/azure/core/serializer/avro/apache/ApacheAvroSerializer.java

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.serializer.avro.apache;
+
+import com.azure.core.serializer.AvroSerializer;
+import com.azure.core.util.logging.ClientLogger;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericDatumWriter;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.DatumWriter;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.io.EncoderFactory;
+import reactor.core.Exceptions;
+import reactor.core.publisher.Mono;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Objects;
+
+/**
+ * Apache based implementation of the {@link AvroSerializer} interface.
+ */
+public class ApacheAvroSerializer implements AvroSerializer {
+    private static final Schema.Parser PARSER = new Schema.Parser();
+    private static final DecoderFactory DECODER_FACTORY = DecoderFactory.get();
+    private static final EncoderFactory ENCODER_FACTORY = EncoderFactory.get();
+
+    private final ClientLogger logger = new ClientLogger(ApacheAvroSerializer.class);
+
+    @Override
+    public <T> Mono<T> read(byte[] input, String schema) {
+        return Mono.fromCallable(() -> {
+            if (input == null) {
+                return null;
+            }
+
+            try {
+                DatumReader<T> reader = new GenericDatumReader<>(PARSER.parse(schema));
+                return reader.read(null, DECODER_FACTORY.binaryDecoder(input, null));
+            } catch (IOException ex) {
+                throw logger.logExceptionAsError(Exceptions.propagate(ex));
```

Comment:
Should the order be reversed? I am not sure which one actually propagates the exception correctly in error channel.
`throw Exceptions.propagate(logger.logExceptionAsError(ex))`